soc/riscv/riscv-ite/it8xxx2/ilm.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2022 The ChromiumOS Authors.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+
+LOG_MODULE_REGISTER(soc_it8xxx2_ilm, CONFIG_LOG_DEFAULT_LEVEL);
+
+/*
+ * Instruction Local Memory (ILM) support for IT8xxx2.
+ *
+ * IT8xxx2 allows 4-kilobyte blocks of RAM be configured individually as either Instruction- or
+ * Data Local Memory (ILM or DLM). Addresses from which instructions will be fetched by the CPU
+ * *must* be in the Flash memory space: it is not permitted to execute from RAM addresses, only
+ * through ILM mappings into RAM.
+ *
+ * When a RAM block is configured as ILM, accesses to addresses matching the corresponding Scratch
+ * SRAM address register (SCARn{H,M,L}) are redirected to the corresponding ILM block in RAM.
+ * If SCAR0 (corresponding to ILM0) has the value 0x8021532 and ILM0 is enabled, then instruction
+ * fetches from the memory range 0x8021532..0x8022532 will be redirected to physical addresses
+ * 0x80100000..0x80101000 (the first 4k block of RAM).
+ *
+ * Instruction fetch from Flash is normally cacheable, but configuring ILM for a region makes that
+ * address range non-cacheable (which is appropriate because Flash has high latency but RAM is
+ * essentially the same speed as cache).
+ */
+
+extern const uint8_t __ilm_flash_start[];
+extern const uint8_t __ilm_flash_end[];
+extern uint8_t __ilm_ram_start[];
+extern uint8_t __ilm_ram_end[];
+
+#define ILM_BLOCK_SIZE 0x1000
+BUILD_ASSERT((ILM_BLOCK_SIZE & (ILM_BLOCK_SIZE - 1)) == 0, "ILM_BLOCK_SIZE must be a power of two");
+
+#define FLASH_BASE CONFIG_FLASH_BASE_ADDRESS
+#define RAM_BASE   CONFIG_SRAM_BASE_ADDRESS
+#define RAM_SIZE   CONFIG_SRAM_SIZE
+
+#define ILM_NODE DT_NODELABEL(ilm)
+
+#define SCARH_ENABLE	 BIT(3)
+#define SCARH_ADDR_BIT19 BIT(7)
+
+/*
+ * SCAR registers contain 20-bit addresses in three registers, with one set
+ * of SCAR registers for each ILM block that may be configured.
+ */
+struct scar_reg {
+	/* Bits 0..7 of address; SCARnL */
+	uint8_t l;
+	/* Bits 8..15 of address; SCARnM */
+	uint8_t m;
+	/* Bits 16..18 and 19 of address, plus the enable bit for the entire SCAR; SCARnH */
+	uint8_t h;
+};
+
+struct ilm_config {
+	/* Points to the contiguous set of RVILMCRn registers */
+	volatile uint8_t *ilm_control_regs;
+	volatile struct scar_reg *scar_regs[15];
+};
+
+bool it8xxx2_is_ilm_configured(void)
+{
+	return device_is_ready(DEVICE_DT_GET(ILM_NODE));
+}
+
+static bool __maybe_unused is_block_aligned(const void *const p)
+{
+	return ((uintptr_t)p & (ILM_BLOCK_SIZE - 1)) == 0;
+}
+
+static int it8xxx2_configure_ilm_block(const struct ilm_config *const config, void *ram_addr,
+				       const void *flash_addr, const size_t copy_sz)
+{
+	if ((uintptr_t)ram_addr < RAM_BASE)
+		return -EFAULT; /* Not in RAM */
```

Comment:
Please wrap all if/loop statements in braces (also see https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style )

Reply:
Done.